tests/AggregationBlogPost/article.md

```diff
@@ -0,0 +1,713 @@
+    Expression\ResolvesToDate $reportDate,
+): Expression\ResolvesToInt {
+    return Expression::min(
+        Expression::dateDiff(
+            startDate: $previousReportDate,
+            endDate: $reportDate,
+            unit: 'second',
+        ),
+        self::computeElapsedSecondsOnDay($reportDate),
+    );
+}
+```
+
+Again, this reduces the complexity of the pipeline stage tremendously:
+
+```php
+$prices = Expression::arrayFieldPath('prices');
+$reportDate = Expression::variable('priceReport.reportDate');
+$previousReportDate = Expression::variable('priceReport.previous.reportDate');
+
+$pipeline = new Pipeline(
+    Stage::addFields(
+        weightedPrices: Expression::map(
+            input: $prices,
+            in: object(
+                duration: self::computeDurationBetweenDates($previousReportDate, $reportDate),
+                price: Expression::variable('priceReport.previous.price'),
+            ),
+            as: 'priceReport',
+        ),
+        lastPrice: Expression::last($prices),
+    ),
+);
+```
+
+### Create Your Own Builder
+
+One last suggestion for reusable code is to create your own builder. When
+extracting complex expressions or stages into separate methods, don't keep those
+internal, but make them public as shown above. For one, this allows reusing them
+in other places, but more importantly, it allows testing them in isolation.
+
+Think about the complete pipeline shown at the beginning: you would most likely
+have tests to ensure you're receiving the correct result, but if that test fails
+you'll be forced to evaluate the entire pipeline to figure out what's wrong.
+This would include all the logic included in the pipeline. By extracting complex
+parts into separate methods, you can test those in isolation under controlled
+conditions, knowing that they behave as you intended. Looking at the last
+example where we've extracted the expression that computes the duration between
+two dates, you can now easily verify that all the individual parts of that
+complex expression work as expected.
+
+Just beware of premature abstractions: extract complex logic with the goal of
+making your code more readable and testable, not to be able to reuse it across
+your entire code base.
+
+## Builder Internals
+
+In the examples you've seen so far, we've only ever looked at stages,
+expressions, and of course the `Pipeline` objects. But how do we end up with an
+aggregation pipeline that the server understands? The objects created with the
+various factory methods are value holder objects. They don't contain any logic,
+but they allow us to know what types an expression will resolve to.
+
+For an example, let's take the `$hour` operator. When you use
+`Expression::hour`, you will receive an instance of an `HourOperator` class.
+This class implements an `OperatorInterface`, telling the builder that this is
+an operator that can be used in aggregation pipeline. It also implements a
```

Comment:
```suggestion
an operator that can be used in an aggregation pipeline. It also implements a
```

Technically, would an operator be used within a pipeline _stage_?

Pipelines have 1+ stages, and stages have some expressions and operators.